### PR TITLE
Three birdanim - resolves #304

### LINKED ISSUE #304: Support for three.js animated birds

### PR DESCRIPTION
**Pull request by kkananen**
support for arrays of geometries and materials
material modifier node

### CODE REVIEW

File: browser/plugins/three_loader_model.plugin.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+		if (this.materials)
+			this.materials = null
+
+		THREE.Loader.Handlers.add(/\.dds$/i, new THREE.DDSLoader())
+
+		console.log('AbstractThreeLoaderObjPlugin loading', this.state.url)
+
+		var url = this.state.url
+		var extname = url.substring(url.lastIndexOf('.'))
```
Comment (kschzt) [inline browser/plugins/three_loader_model.plugin.js]: extra var

File: browser/plugins/three_material_modifier.plugin.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+		this.material_override = {}
+
+		this.material_slots_dirty = false
+		this.materials_dirty = false
+	}
+
+	ThreeMaterialModifier.prototype.update_input = function(slot, data) {
+		if (!slot.uid && slot.index == 0) {
+			if (data) {
+				if (this.in_material_array.length == data.length) {
```
Comment (kschzt) [inline browser/plugins/three_material_modifier.plugin.js]: triple equals, and early out if !data

File: browser/plugins/three_material_modifier.plugin.js [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+					for (var i = 0; i < this.in_material_array.length; ++i) {
+						if (this.in_material_array[i] != data[i]) {
```
Comment (kschzt) [inline browser/plugins/three_material_modifier.plugin.js]: triple equals

File: browser/plugins/three_material_modifier.plugin.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+							break
+						}
+					}
+
+					if (i == this.in_material_array.length) {
```
Comment (kschzt) [inline browser/plugins/three_material_modifier.plugin.js]: triple equals

File: browser/plugins/three_material_modifier.plugin.js [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+				this.in_material_array = data
+				this.out_material_array = []
+				this.material_override = {}
+
+				this.material_slots_dirty = true
+			}
+		}
+		else { // dynamic slot
+			if (!data) {
+				if (slot.name in this.material_override) {
```
Comment (kschzt) [inline browser/plugins/three_material_modifier.plugin.js]: you should write this as `this.material_override.indexOf(slot.name) !== -1`

File: browser/scripts/abstractThreeLoaderObjPlugin.js [hunk 6]
```diff
@@ -40,11 +55,19 @@ AbstractThreeLoaderObjPlugin.prototype.create_ui = function() {
 }
 
 AbstractThreeLoaderObjPlugin.prototype.update_input = function(slot, data) {
-	if (slot.uid) {
+	/*if (slot.uid) {
 		this.materials[slot.name] = data
 
-		if (!this.childrenByMaterialName[slot.name])
+		if (!this.childrenByMaterialName[slot.name]) {
```
Comment (kschzt) [inline browser/scripts/abstractThreeLoaderObjPlugin.js]: pls remove commented code

File: browser/scripts/abstractThreeLoaderObjPlugin.js [hunk 7]
```diff
@@ -85,31 +109,42 @@ AbstractThreeLoaderObjPlugin.prototype.adjustMaterialSlots = function() {
...
-			that.childrenByMaterialName[matName].push(child)
+			this.childrenByMaterialName[matName].push(geom)
 		}
-	})
+	}
 
-	this.adjustMaterialSlots()
+	if (this.pendingMaterialInput) {
+		for (input in Object.keys(this.pendingMaterialInput)) {
+			for (var i=0; i < this.childrenByMaterialName[input].length; i++)
```
Comment (kschzt) [inline browser/scripts/abstractThreeLoaderObjPlugin.js]: remove commented code

**Comment by kschzt** [comment]
otherwise looks good - please squash the commits before merging

File: browser/plugins/three_material_modifier.plugin.js (hunk 2 above)
Comment (kkananen) [inline browser/plugins/three_material_modifier.plugin.js]: can't early out on !data as we need to remove stuff from the array in the else block

**Pull request merged**